p4-16/spec/P4-16-spec.mdk

```diff
@@ -6416,6 +6416,153 @@ ck16.update( { h.ipv4.ttl, h.ipv4.proto } );
+|             | can be a constructor parameter for this type ||||
+|             |---------|--------|---------|--------|
+| This type   | package | parser | control | extern |
++-------------+:-------:+:------:+:-------:+:------:+
+| package     | yes     |  no    |  no     |  no    |
+| parser      | yes     | yes    |  no     |  no    |
+| control     | yes     |  no    | yes     |  no    |
+| extern      | yes     | yes    | yes     | yes    |
+| table       |  no     |  no    |  no     |  no    |
+| value types | yes     | yes    | yes     | yes    |
+|-------------|---------|--------|---------|--------|
+
+The next table lists restrictions on where one may perform
+instantiations (see Section [#sec-instantiations]) of different types.
+The answer for `package` is always "no" because there is no
+"inside a package" where instantiations can be written in P4~16~.  One
+can definitely make constructor calls and use instances of stateful
+types as parameters when instantiating a package, and restrictions on
+those types are in the table above.
+
+For externs, one can only specify their interface in P4~16~, not their
+implementation.  Thus there is no place to instantiate objects within
+an extern.
+
+You may declare variables and constants of any of the value types
+within a parser or control.  Declaring a variable or constant is not
+the same as instantiation, hence the answer "no" in those table
+entries.  Variables may not be declared at the top level of your
+program, but constants may.
+
+|-------------|-------------------------------------------------|
+|             | can be instantiated in this place |||||
+|             |-----------|---------|---------|---------|--------|
+| This type   | top level | package | parser  | control | extern |
++-------------+:---------:+:-------:+:-------:+:-------:+:------:+
+| package     | yes       |  no     | no      | no      |  no    |
+| parser      |  no       |  no     | yes     | no      |  no    |
+| control     |  no       |  no     | no      | yes     |  no    |
+| extern      | yes       |  no     | yes     | yes     |  no    |
+| table       |  no       |  no     | no      | yes     |  no    |
+| value types |  no       |  no     | no      | no      |  no    |
+|-------------|-----------|---------|---------|---------|--------|
+
+The next table lists restrictions on what types can be passed as
+run-time parameters to other callable things that have run-time
+parameters: parsers, controls, extern methods, and actions.
+
+|-------------|-------------------------------------------------|
+|             | can be a run-time parameter to this callable thing ||||
+|             |-----------|---------|---------|---------|
+| This type   | parser    | control | method  | action  |
++-------------+:---------:+:-------:+:-------:+:-------:+
+| package     |  no       |  no     | no      | no      |
+| parser      |  no       |  no     | no      | no      |
```

Comment:
We already support calling subparsers from parsers. If we want to support modularity, we have to allow that someone writes a parser in a separate module and passes that instance to the "main" parser. I can't think of any other method by which existing written code can take a not-yet-seen parser, except if it can be passed at runtime.